### PR TITLE
Add TTSsay squirrel function

### PR DESCRIPTION
**Pull request by barnabwhy**
:trollface: client mods which say things are now possible
example usage:
```cpp
TTSsay("BobTheBob9")
```

### CODE REVIEW

File: NorthstarDedicatedTest/voicebasic.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+  public:
+	float speechRate;
+	BasicVoice()
+	{
+		pVoice = NULL; /* The ISpVoice interface enables an application to perform text synthesis operations.
+		Applications can speak text strings and text files, or play audio files through this interface.
+		All of these can be done synchronously or asynchronously.*/
+		input = L"";
+		a = CoInitializeEx(NULL, COINIT_APARTMENTTHREADED); // HRESULT CoInitializeEx(LPVOID pvReserved, DWORD dwCoInit);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:35]: This doesn't belong in a constructor.

File: NorthstarDedicatedTest/voicebasic.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  protected:
+	int choice;
+	ISpVoice* pVoice;
+	HRESULT hr, a; /*In the field of computer programming, the HRESULT is a data type used in Windows operating systems, and the earlier
+IBM/Microsoft OS/2 operating system, to represent error conditions, and warning conditions. The original purpose of HRESULTs was to formally
+lay out ranges of error codes for both public and Microsoft internal use in order to prevent collisions between error codes in different
+subsystems of the OS/2 operating system. HRESULTs are numerical error codes. Various bits within an HRESULT encode information about the
+nature of the error code, and where it came from.
+HRESULT error codes are most commonly encountered in COM programming, where they form the basis for a standardized COM error handling
+convention.*/
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:17]: Seriously?

File: NorthstarDedicatedTest/voicebasic.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		You will access each COM object from a single thread; you will not share COM interface pointers between multiple threads.
+		The thread will have a message loop
+		Diff b/w appartment and multi threading
+		https://stackoverflow.com/questions/485086/single-threaded-apartments-vs-multi-threaded-apartments*/
+		if (FAILED(a))
+		{
+			cout << "ERROR 404 FAILED INITIALIZING COM\n";
+			exit(1);
+		}
+		hr = CoCreateInstance(CLSID_SpVoice, NULL, CLSCTX_ALL, IID_ISpVoice, (void**)&pVoice);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:49]: You haven't checked errors for this.

File: NorthstarDedicatedTest/voicebasic.h (hunk 2 above)
Comment (barnabwhy) [reply NorthstarDedicatedTest/voicebasic.h:17]: I copied a lot of code :)

Comment (emma-miler) [reply NorthstarDedicatedTest/voicebasic.h:17]: don't we all :P

File: NorthstarDedicatedTest/voicebasic.h (hunk 1 above)
Comment (barnabwhy) [reply NorthstarDedicatedTest/voicebasic.h:35]: Why not?

File: NorthstarDedicatedTest/voicebasic.h (hunk 2 above)
Comment (barnabwhy) [reply NorthstarDedicatedTest/voicebasic.h]: No more comments

File: NorthstarDedicatedTest/voicebasic.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+void BasicVoice::speakText(char* input)
+{
+	size_t size = strlen(input) + 1;
+	wchar_t* wtext = new wchar_t[size];
+
+	size_t outSize;
+	mbstowcs_s(&outSize, wtext, size, input, size - 1);
+	LPWSTR ptr = wtext;
+	hr = pVoice->Speak(ptr, SVSFlagsAsync + SVSFPurgeBeforeSpeak, NULL);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.cpp:12]: Check if pVoice is null first.

File: NorthstarDedicatedTest/voicebasic.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+void BasicVoice::setRate(float rate)
+{
+	speechRate = rate;
+	pVoice->SetRate(rate);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.cpp:18]: ^

File: NorthstarDedicatedTest/voicebasic.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+}
+void BasicVoice::setVolume(float volume)
+{
+	speechVolume = volume;
+	pVoice->SetVolume(volume * 100.0);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.cpp:23]: ^

File: NorthstarDedicatedTest/voicebasic.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	}
+	virtual void speakText(char* input);
+	virtual void setRate(float rate);
+	virtual void setVolume(float volume);
+	virtual void outSpeech();
+	virtual void skipSpeech();
+	virtual ~BasicVoice()
+	{
+		::CoUninitialize();
+		delete pVoice;
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:32]: Why are you deleting pVoice? COM objects are managed with the reference counting from the release function.

File: NorthstarDedicatedTest/voicebasic.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+void BasicVoice::outSpeech()
+{
+	pVoice->Release();
+	pVoice = NULL;
+	::CoUninitialize();
+}
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.cpp:31]: What is the point of this method? It's not called anywhere, and it looks like it should be the destructor.

File: NorthstarDedicatedTest/voicebasic.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+void BasicVoice::skipSpeech()
+{
+	unsigned long skipped;
+	pVoice->Skip(L"SENTENCE", 1, &skipped);
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.cpp:36]: ...

File: NorthstarDedicatedTest/voicebasic.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  protected:
+	ISpVoice* pVoice;
+	HRESULT hr, a;
+
+  public:
+	float speechRate = 1.0;
+	float speechVolume = 1.0;
+	BasicVoice()
+	{
+		pVoice = NULL;
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:17]: This is unnecessary. Class members are zero-initialized by default.

**Review by pg9182 — CHANGES_REQUESTED** [review]

File: NorthstarDedicatedTest/voicebasic.cpp (hunk 8 above)
Comment (barnabwhy) [reply NorthstarDedicatedTest/voicebasic.cpp:31]: It's a remnant of code that did call it, will sort that out

**Comment by abarichello** [comment]
Apparently SAPI doesn't work on wine so I can't test this

**Comment by abarichello** [comment]
needs some conflict solving btw

File: NorthstarDedicatedTest/voicebasic.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+		{
+			hr = CoCreateInstance(CLSID_SpVoice, NULL, CLSCTX_ALL, IID_ISpVoice, (void**)&pVoice);
+		}
+	}
+	virtual void speakText(char* input);
+	virtual void setRate(float rate);
+	virtual void setVolume(float volume);
+	virtual ~BasicVoice()
+	{
+		pVoice->Release();
```
Comment (pg9182) [inline NorthstarDedicatedTest/voicebasic.h:29]: You don't check if pVoice is NULL here.

**Comment by GeckoEidechse** [comment]
Closing as stale. We might still want to do this for accessibility reasons but unless someone mentions they actually need it, it's pretty low on our todo list I guess.

Feel free to re-open with requested changes applied.

**Pull request closed without merging**